File: src/Main/lua/ga/corelauncher/Libraries/PipeObject.lua
local function RandomString(Length)
    local Characters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
    local String = ""
    for i = 1, Length do
        local RandomIndex = math.random(1, #Characters)
        String = String .. string.sub(Characters, RandomIndex, RandomIndex)
    end
    return String
end

local function GetObjectFunctions(Object)
    local Functions = {}

    for Key, Value in pairs(TypeWriter.JavaScript.Global.Object:getOwnPropertyNames(TypeWriter.JavaScript.Global.Object:getPrototypeOf(Object))) do
        if Value ~= "constructor" and Value ~= "length" then
            Functions[Value] = Object[Value]
        end
    end

    return Functions
end

local function CloneWithoutFunctions(Object)
    if type(Object) ~= "table" and TypeWriter.JavaScript.TypeOf(Object) ~= "object" then
        return Object
    end

    if Object == nil then
        return nil
    end

    local Cloned = {}
    for Key, Value in pairs(Object) do
        if type(Value) == "function" or TypeWriter.JavaScript.TypeOf(Value) == "function" or Key == "length" then
        
        elseif type(Value) == "table" or TypeWriter.JavaScript.TypeOf(Value) == "object" then
            Cloned[Key] = CloneWithoutFunctions(Value)
        else
            Cloned[Key] = Value
        end
    end

    return Cloned
end

local PipeCache = {}
local function PipeObject(Name, Object)
    local PipedObject = {}

    for Key, Value in pairs(GetObjectFunctions(Object)) do
        if PipeCache[Value] then
            PipedObject[Key] = PipeCache[Value]
        else
            local PipeHandle = "PipeFunction." .. Name .. "." .. Key .. "." .. RandomString(16)
    
            CoreLauncher.IPCMain.handle(
                PipeHandle,
                function(...)
                    local Arguments = { ... }
                    table.remove(Arguments, 1)
                    return CloneWithoutFunctions(Value(table.unpack(Arguments)))
                end
            )
    
            PipedObject[Key] = { PipeType = "function", PipeHandle = PipeHandle }
            PipeCache[Value] = PipedObject[Key]
        end
    end

    return PipedObject
end

return PipeObject

File: src/Main/js/ga/corelauncher/Libraries/PipeObject.js
const RandomString = require("randomstring").generate

function GetObjectFunctions(Obj) {
    const Functions = {}

    const Prototype = Object.getPrototypeOf(Obj)
    const PropertyNames = Object.getOwnPropertyNames(Prototype)
    for (const FunctionName of PropertyNames) {
        Functions[FunctionName] = Obj[FunctionName]
    }

    return Functions
}

function CloneWithoutFunctions(Object) {
    if (typeof Object != "object") { return Object }

    if (Object instanceof Array) {
        const NewArray = []
        for (const Value of Object) {
            if (typeof Value == "function") { continue }
            NewArray.push(CloneWithoutFunctions(Value))
        }
        return NewArray
    } else {
        const NewObject = {}
        for (const Key in Object) {
            const Value = Object[Key]
            if (typeof Value == "function") { continue }
            NewObject[Key] = CloneWithoutFunctions(Value)
        }
        return NewObject
    }
}

var PipeCache = {}
function PipeObject(Name, Object) {
    var PipedObject = {}

    var Functions = GetObjectFunctions(Object)
    for (const Key in Functions) {
        const Value = Functions[Key]
        if (PipeCache[Value]) {
            PipedObject[Key] = PipeCache[Value]
        } else {
            const PipeHandle = `PipeFunction.${Name}.${Key}.${RandomString(16)}`

            CoreLauncher.IPCMain.handle(
                PipeHandle,
                async function () {
                    const Arguments = Array.prototype.slice.call(arguments)
                    Arguments.shift()
                    return CloneWithoutFunctions(await Value.apply(Object, Arguments))
                }
            )

            PipedObject[Key] = { PipeType: "function", PipeHandle: PipeHandle }
            PipeCache[Value] = PipedObject[Key]
        }
    }

    return PipedObject
}

module.exports = PipeObject
